src/tvp/model.ts

```diff
@@ -0,0 +1,352 @@
+    }
+  }
+
+  public async height(): Promise<number> {
+    if (this.isExpanded()) {
+      const children = await this.getChildren()
+      const heights = await Promise.all(children.map(child => child.height()))
+      return heights.reduce((h1, h2) => h1 + h2, 1)
+    } else {
+      return 1
+    }
+  }
+
+  private async collectVisibleNodes(): Promise<NodeView[]> {
+    async function collectInternal(acc: NodeView[], node: Node): Promise<NodeView[]> {
+      const newAcc = acc.concat(node.makeView())
+      if (node.isExpanded()) {
+        const children = await node.getChildren()
+        return children.reduce((zP, child) => zP.then(z => collectInternal(z, child)), Promise.resolve(newAcc))
+      } else {
+        return newAcc
+      }
+    }
+    return collectInternal([], this)
+  }
+
+  private async touch(): Promise<number> {
+    const state = this.lastTouch = Date.now()
+    if (this.isExpanded()) {
+      const children = await this.getChildren()
+      return Promise.all(children.map(child => child.touch())).then(_ => state)
+    } else {
+      return state
+    }
+  }
+
+  private async markAsDefunct(): Promise<void> {
+    this.defunct = true
+    if (this.isExpanded()) {
+      const children = await this.getChildren()
+      await Promise.all(children.map(child => child.markAsDefunct()))
+    }
+  }
+
+  public makeView(): NodeView {
+    return {
+      underlying: this.viewNode,
+      level: this.level,
+      expandable: this.expandable(),
+      expanded: this.isExpanded()
+    }
+  }
+}
+
+export class TreeModel implements Disposable {
+  private emitter = new Emitter<TreeModelUpdate>()
+  public readonly updateEvents = this.emitter.event
+  public readonly rootNode: Node
+  private visible: boolean = false
+
+  constructor(
+    public provider: TreeViewProvider,
+  ) {
+    const viewId = provider.viewId
+    const treeViewNode: TreeViewNode = { viewId, label: viewId }
+    this.rootNode = new Node(treeViewNode, 0, provider, this.emitter)
+    provider.updatedNodes(node => {
+      if (node.nodeUri !== undefined) {
+        void this.findNodeByUri(node.nodeUri).then(treeNode => {
+          if (treeNode !== undefined) return treeNode.refreshSubtree(node)
+        })
+      } else {
+        void this.rootNode.refreshSubtree(undefined)
+      }
+    })
+  }
+
+  public get viewId(): string {
+    return this.provider.viewId
+  }
+
+  public show(): void {
+    if (!this.visible) {
+      this.visible = true
+      return this.provider.sendTreeViewVisibilityNotification(true)
+    }
+  }
+
+  public hide(): void {
+    if (this.visible) {
+      this.visible = false
+      return this.provider.sendTreeViewVisibilityNotification(false)
+    }
+  }
+
+  private async findNodeWithOffsetInternal(node: Node, offset: number): Promise<Node | undefined> {
+    if (offset == 0) {
+      return node
+    } else {
+      const children = await node.getChildren()
+      return children
+        .reduce((stateP, child) => {
```

Comment:
Can we change the variable name `stateP`? Not a huge fan of ending with a cap letter, and also a fan of more explicit naming even if it's a bit more verbose.